### PR TITLE
Add RoBERTa FP8 support with refactoring

### PR DESCRIPTION
**Pull request by andrea-fasoli**
This PR introduces launch scripts and utility functions for FP8 encoders inference: RoBERTa for QuestionAnswering or MaskedLM architecture and tasks are supported.

The PR incorporates several aspects of the ongoing refactoring process for INT8 LLM from #34 (including addressing the issues raises during that review process), but limits updates to encoder models (i.e., no changes to LLM nor `inference.py`) and focuses on FP8 quantization instead. It assumes support for FP8 has been incorporated in FMS and FMS-MO (ongoing effort).

Main launch script for encoders is `scripts/encoders_inference.py`
Utility for various setup steps and inference are:
```
aiu_setup               set up AIU environment variables
args_parsing            define script arguments across all model configurations
encoders_utils          classes to run QA and MaskedLM tasks
model_setup             define model dtype, device, and distributed strategy
quantization_setup      import FMS-MO addons and define linear_config if needed
```

### CODE REVIEW

**Comment by andrea-fasoli** [comment]
cc: @ani300 @JRosenkranz

File: aiu_fms_testing_utils/utils/args_parsing.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        help="size of input batch",
+    )
+    args_dec_enc.add_argument(
+        "--max_prompt_length",
+        type=int,
+        default=None,
+        help=(
+            "Cap the number of tokens per prompt to a maximum length prior to padding. "
+            "If None, prompts to decoder models will have no cap, while prompts to "
+            "encoder models will be capped to a default of 384 tokens."
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/args_parsing.py:163]: where is this limit of 384 in the code?

File: aiu_fms_testing_utils/utils/args_parsing.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        "--distributed",
+        action="store_true",
+        help="This is a distributed job (multiple instances run with RANK+WORLD_SIZE)",
+    )
+    args_run_settings.add_argument(  # could be a bool / flag
+        '-v', '--verbose',
+        action='count',
+        default=0,
+        help="Set verbosity level (pass flag as `-v`, `-vv`, `-vvv`)"
+    )
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/args_parsing.py:114]: Why are we making different levels of verbosity instead of a flag given the code usecases?

File: aiu_fms_testing_utils/utils/args_parsing.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+    # Quantization arguments
+    args_quantization = parser.add_argument_group("Model quantization")
+    args_quantization.add_argument(
+        "--quantization",
+        type=str,
+        choices=["gptq", "fp8"],  # TODO: add "fp8" when available in FMS
+        default=None,
+        help="Type of quantization of the model checkpoint",
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/args_parsing.py:61]: remove TODO, and add mention that can also be inferred from checkpoint

File: aiu_fms_testing_utils/utils/args_parsing.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        "--max_eval_samples",
+        type=int,
+        default=None,
+        help=(
+            "For debugging purposes or quicker training, truncate the number of "
+            "evaluation examples to this value if set."
+        ),
+    )
+    args_encoder.add_argument(
+        "--preprocessing_num_workers", type=int, default=1, help=""
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/args_parsing.py:301]: add a help comment

File: aiu_fms_testing_utils/utils/args_parsing.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+        type=str,
+        default=None,
+        help="A csv or a json file containing the validation data.",
+    )
+    args_encoder.add_argument(
+        "--pad_to_max_length",
+        action="store_true",
+        help=(
+            "If passed, pad all samples to `max_seq_length`. "
+            "Otherwise, dynamic padding is used."
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/args_parsing.py:288]: what do you mean by dynamic padding here?

File: aiu_fms_testing_utils/utils/quantization_setup.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            group_size = 128
+            desc_act = False
+
+        linear_config = {
+            "linear_type": linear_type,
+            "group_size": group_size,
+            "desc_act": desc_act,
+        }
+    elif args.quantization == "fp8":
+        dprint("fp8 config is inferred from HF checkpoint via FMS / FMS-MO functions")
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/quantization_setup.py:82]: maybe mark this as [INFO]

File: aiu_fms_testing_utils/utils/model_setup.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    else:
+        device = torch.device(args.device_type)
+    return device
+
+
+def print_system_setup(args: argparse.Namespace) -> None:
+    """Display system info (rank 0 only)."""
+
+    if rank == 0 and args.verbose:
+        dprint("-"*60)
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/model_setup.py:65]: no need to use dprint if you already cap the print to rank 0

File: scripts/encoders_inference.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    distributed.init_process_group(backend="gloo", rank=rank, world_size=world_size)
+
+# Main model setup
+default_dtype, device, dist_strat = setup_model(args)
+
+# Retrieve linear configuration (quantized or not) to instantiate FMS model
+linear_config = get_linear_config(args)
+
+if rank == 0:
+    dprint("="*60)
```
Comment (ani300) [inline scripts/encoders_inference.py:44]: rank0 and dprint not needed

File: aiu_fms_testing_utils/utils/encoders_utils.py [hunk 9]
```diff
@@ -0,0 +1,723 @@
...
+        """FMS uses a different standard than HF for encoder inputs."""
+
+        return {'x': batch['input_ids'], 'mask': batch['attention_mask']}
+
+    def process_eval_set(self) -> None:
+        """Pre-process evaluation dataset for QuestionAnswering task."""
+
+        if not has_hf:
+            raise ImportError(
+                "QuestionAnswering Encoder requires transformer package but import "
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/encoders_utils.py:177]: typo: transformer -> transformers

File: aiu_fms_testing_utils/utils/encoders_utils.py [hunk 10]
```diff
@@ -0,0 +1,723 @@
...
+
+        args = self.args
+
+        q_col_name = self.question_column_name
+        c_col_name = self.context_column_name
+        pad_on_right = self.pad_on_right
+        max_prompt_length = (
+            args.max_prompt_length
+            if args.max_prompt_length is not None
+            else 384
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/encoders_utils.py:110]: should we get this from the model itself?

File: aiu_fms_testing_utils/utils/args_parsing.py (hunk 1 above)
Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/args_parsing.py:163]: `utils/encoder_utils.py` --> class `EncoderQAInfer` --> `prepare_validation_features`

File: aiu_fms_testing_utils/utils/encoders_utils.py [hunk 11]
```diff
@@ -0,0 +1,723 @@
...
+            # During Feature creation dataset samples might increase, we will select
+            # required samples again
+            eval_dataset = eval_dataset.select(range(args.max_eval_samples))
+
+        # store evaluation dataset prior dropping
+        self.eval_dataset = eval_dataset
+
+        # DataLoaders creation:
+        if args.pad_to_max_length:
+            # If padding was already done ot max length, we use the default data collator
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/encoders_utils.py:247]: typo: ot -> to

File: aiu_fms_testing_utils/utils/args_parsing.py (hunk 2 above)
Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/args_parsing.py:114]: we don't use verbosity levels at this time. Changed to flag.

File: aiu_fms_testing_utils/utils/encoders_utils.py [hunk 12]
```diff
@@ -0,0 +1,723 @@
...
+                    all_predictions[example["id"]] = best_non_null_pred["text"]
+
+            # Make `predictions` JSON-serializable by casting np.float back to float.
+            all_nbest_json[example["id"]] = [
+                {k: (float(v) if isinstance(v, (np.float16, np.float32, np.float64)) else v) for k, v in pred.items()}
+                for pred in predictions
+            ]
+
+        # If we have an output_dir, let's save all those dicts.
+        if output_dir is not None:
```
[the remark this answers is not in the document]
Comment (ani300) [reply aiu_fms_testing_utils/utils/encoders_utils.py:466]: I guess all preprocessing really

File: aiu_fms_testing_utils/utils/args_parsing.py (hunk 3 above)
Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/args_parsing.py:61]: TODO removed.

inferring quantization from ckpt works at FMS level but is not compatible with these scripts. `is_quantized` boolean is initialized based on whether this `quantization` argument is not None. `is_quantized` is used in `encoders_inference.py` to import addons before `get_model` from FMS is called.

File: aiu_fms_testing_utils/utils/encoders_utils.py [hunk 13]
```diff
@@ -0,0 +1,723 @@
...
+            if step + batch_size < len(dataset):
+                logits_concat[step : step + batch_size, :cols] = output_logit
+            else:
+                logits_concat[step:, :cols] = output_logit[: len(dataset) - step]
+
+            step += batch_size
+
+        return logits_concat
+
+    def run_warmup(self) -> None:
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/encoders_utils.py:573]: If we aren't measuring running times, we don't need warmup for AIU encoder models

File: scripts/encoders_inference.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+tokenizer = tokenizers.get_tokenizer(args.tokenizer)
+
+model.eval()
+set_grad_enabled(False)
+if args.distributed:
+    distributed.barrier()
+dprint(f"Loading model completed in {time.time() - loading_model_start:.2f} s.")
+
+if isinstance(model, RoBERTa):
+    model = wrap_encoder(model)  # enable using pipeline to eval RoBERTa MaskedLM
```
Comment (ani300) [inline scripts/encoders_inference.py:78]: I have a question on the task really requiring the pipeline

**Review by ani300 — COMMENTED** [review]
I have left a few comments, but I'm optimistic about merging this tomorrow

**Comment by ani300** [comment]
Can we also remove the current roberta example script in FMS? I think this supersedes it.

File: aiu_fms_testing_utils/utils/args_parsing.py (hunk 5 above)
Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/args_parsing.py:288]: if `pad_to_max_length` is passed, the batches in the evaluation dataloader are padded to the selected length.
if None, we pad to the max sequence of each batch. The latter is the meaning of "dynamic padding" in the torch example we used as basis for QA. I clarified the help, removing mention of dynamic padding which can be confusing.

Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/args_parsing.py:288]: I don't know about minimum pad size or specific prompt lengths to meet for encoders. Default is fixed length of 384 currently. Should be easy to implement other constraints now or at a later date. I reached out to the compiler team for info.

File: aiu_fms_testing_utils/utils/encoders_utils.py (hunk 13 above)
Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/encoders_utils.py:573]: would it be a functionality that we may want to have down the line, if the focus shifts back to encoders? I think we could keep this

File: aiu_fms_testing_utils/utils/encoders_utils.py [hunk 15]
```diff
@@ -0,0 +1,723 @@
...
+        """
+
+        dprint(f"Starting evaluation ({warmup=})...")
+        warmup_start_time = time.time()
+        unmasker = pipeline(
+            "fill-mask",
+            model=self.model,
+            tokenizer=self.tokenizer.tokenizer,
+        )
+        output = unmasker(self.prompt)
```
[the remark this answers is not in the document]
Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/encoders_utils.py:684]: answered below

File: aiu_fms_testing_utils/utils/aiu_setup.py [hunk 16]
```diff
@@ -54,11 +56,9 @@ def aiu_setup(rank=0, world_size=1, local_rank=0, local_size=1, verbose=False):
 # ==============================================================
 # Distributed setup
 # ==============================================================
-def aiu_dist_setup(rank, world_size, local_rank=-0, local_size=-1, verbose=False):
+def aiu_dist_setup(rank, world_size, local_rank=-0, verbose=False):
     if local_rank < 0:
         local_rank = rank
-    if local_size < 0:
-        local_size = world_size
```
[the remark this answers is not in the document]
Comment (andrea-fasoli) [reply aiu_fms_testing_utils/utils/aiu_setup.py:61]: good point to keep in mind. Should we add it back at that time or now?

**Comment by andrea-fasoli** [comment]
do we like `encoders_inference.py` as a name for the entry point script?

**Comment by andrea-fasoli** [comment]
do we need any updates to `tests/models/test_encoder.py`?

File: aiu_fms_testing_utils/utils/model_setup.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        if args.device_type == 'aiu':
+            for peer_rank in range(aiu_setup.world_size):
+                pcie_env_str="AIU_WORLD_RANK_"+str(peer_rank)
+                dprint(f"PCI Addr. for Rank {peer_rank} : {os.environ[pcie_env_str]}")
+        print("-"*60)
+
+
+def set_determinism(args: argparse.Namespace) -> None:
+    """Set determinism.
+    NOTE: torch determinism requires env variable: `CUBLAS_WORKSPACE_CONFIG=:4096:8`
```
[the remark this answers is not in the document]
Comment (ani300) [reply aiu_fms_testing_utils/utils/model_setup.py:87]: no, what this really does is ensure cuBLAS always picks the same kernel when doing matmuls on CUDA, as otherwise it'll use some stochastic heuristic which may change results slightly. There are many options, but we just picked this as a single matmul kernel to ensure repetability

File: aiu_fms_testing_utils/utils/encoders_utils.py (hunk 13 above)
Comment (ani300) [reply aiu_fms_testing_utils/utils/encoders_utils.py:560]: sure, we can keep it

**Comment by ani300** [comment]
> do we like encoders_inference.py as a name for the entry point script?

I like `run_encoders.py` better, and it matches `test_encoders`

> the pre- and post-processing functions of the QA task are based on torch QA example. I am sure they can be streamlined (as a non-urgent item). If there's any concern with this, let me know.

Preprocessing needs to run on every process if multi-AIU, but the post-processing should only run on a single rank (unless we want to ensure all ranks get the same output?) Maybe keep it as is as it makes it easier to debug.

> do we need any updates to tests/models/test_encoder.py?

I think the tests do what they need already, which is to ensure the model outputs stay consistent over time using the model signatures framework (which comes from FMS). I don't know if aiu-fms-testing-utils is the place to have validation/e2e tests for specific tasks, @dpatel-ops what's your opinion?

**Review by ani300 — APPROVED** [review]
lgtm!
